src/main/java/mekanism/common/integration/framedblocks/ChemicalCamoContainerFactory.java

```diff
@@ -0,0 +1,143 @@
+package mekanism.common.integration.framedblocks;
+
+import com.mojang.serialization.MapCodec;
+import mekanism.api.Action;
+import mekanism.api.MekanismAPI;
+import mekanism.api.chemical.Chemical;
+import mekanism.api.chemical.ChemicalStack;
+import mekanism.api.chemical.IChemicalHandler;
+import mekanism.api.text.TextComponentUtil;
+import mekanism.common.MekanismLang;
+import mekanism.common.capabilities.Capabilities;
+import net.minecraft.core.BlockPos;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.RegistryFriendlyByteBuf;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import org.jetbrains.annotations.Nullable;
+import xfacthd.framedblocks.api.camo.CamoContainerFactory;
+import xfacthd.framedblocks.api.camo.TriggerRegistrar;
+import xfacthd.framedblocks.api.util.CamoMessageVerbosity;
+import xfacthd.framedblocks.api.util.ConfigView;
+
+final class ChemicalCamoContainerFactory extends CamoContainerFactory<ChemicalCamoContainer> {
+
+    private static final MapCodec<ChemicalCamoContainer> CODEC = Chemical.CODEC
+            .xmap(ChemicalCamoContainer::new, ChemicalCamoContainer::getChemical)
+            .fieldOf("chemical");
+    private static final StreamCodec<RegistryFriendlyByteBuf, ChemicalCamoContainer> STREAM_CODEC =
+            Chemical.STREAM_CODEC.map(ChemicalCamoContainer::new, ChemicalCamoContainer::getChemical);
+    private static final Component MSG_RADIOACTIVE = TextComponentUtil.translate(
+            MekanismLang.FRAMEDBLOCKS_CAMO_RADIOACTIVE.getTranslationKey()
+    );
+
+    @Override
+    protected void writeToNetwork(CompoundTag tag, ChemicalCamoContainer camo) {
+        Chemical chemical = camo.getChemical();
+        tag.putInt("chemical", MekanismAPI.CHEMICAL_REGISTRY.getId(chemical));
+    }
+
+    @Override
+    protected ChemicalCamoContainer readFromNetwork(CompoundTag tag) {
+        Chemical chemical = MekanismAPI.CHEMICAL_REGISTRY.byId(tag.getInt("chemical"));
+        return new ChemicalCamoContainer(chemical);
+    }
+
+    @Override
+    @Nullable
+    public ChemicalCamoContainer applyCamo(Level level, BlockPos pos, Player player, ItemStack stack) {
+        IChemicalHandler handler = Capabilities.CHEMICAL.getCapability(stack);
+        if (handler == null || handler.getChemicalTanks() <= 0) {
+             return null;
+        }
+
+        ChemicalStack chemical = handler.getChemicalInTank(0);
+        if (!isValidChemical(chemical.getChemical(), player)) {
+            return null;
+        }
```

Comment:
Realistically we might want this to check all chemical tanks, and just take the first chemical tank that has a valid chemical in it.

Reply:
Probably keep searching, because if there is say something in an input tank and output and the input is first, it may fail when the output would succeed

Reply:
That's a good point, yeah. No additional change needed then, that's already what the latest code does.